plugins/askrene/algorithm.c

```diff
@@ -0,0 +1,79 @@
+#include "config.h"
+#include <ccan/bitmap/bitmap.h>
+#include <ccan/lqueue/lqueue.h>
+#include <ccan/tal/tal.h>
+#include <plugins/askrene/algorithm.h>
+#include <plugins/askrene/priorityqueue.h>
+
+#define MAX(x, y) (((x) > (y)) ? (x) : (y))
+#define MIN(x, y) (((x) < (y)) ? (x) : (y))
+
+/* Simple queue to traverse the network. */
+struct queue_data {
+	u32 idx;
+	struct lqueue_link ql;
+};
+
+bool BFS_path(const tal_t *ctx, const struct graph *graph,
+	      const struct node source, const struct node destination,
+	      const s64 *capacity, const s64 cap_threshold, struct arc *prev)
+{
+	tal_t *this_ctx = tal(ctx, tal_t);
+	bool target_found = false;
+	const size_t max_num_arcs = graph_max_num_arcs(graph);
+	const size_t max_num_nodes = graph_max_num_nodes(graph);
+
+	/* check preconditions */
+	if (!graph || source.idx >= max_num_nodes || !capacity || !prev)
+		goto finish;
+
+	if (tal_count(capacity) != max_num_arcs ||
+	    tal_count(prev) != max_num_nodes)
+		goto finish;
+
+	for (size_t i = 0; i < max_num_nodes; i++)
+		prev[i].idx = INVALID_INDEX;
+
+	LQUEUE(struct queue_data, ql) myqueue = LQUEUE_INIT;
+	struct queue_data *qdata;
+
+	qdata = tal(this_ctx, struct queue_data);
+	qdata->idx = source.idx;
+	lqueue_enqueue(&myqueue, qdata);
```

Comment:
This reverts an important optimization made in 9f0c0e1ccab9283799a47f021779fe45bb30b167 ?

Reply:
d24e60a643e377a91163afb0271d42958b04d027